pkg/controller/template/template_controller.go

```diff
@@ -362,6 +367,10 @@ func (ctrl *Controller) syncControllerConfig(key string) error {
 		}
 		if updated {
 			glog.V(4).Infof("Machineconfig %s was updated", mcs[idx].Name)
+			if ctrl.readyFlag != nil {
+				ctrl.readyFlag.Done()
+				ctrl.readyFlag = nil
```

Comment:
I believe it's really dangerous to do this, there are many code paths which could lead to Done() not being called causing a deadlock in the goroutines. What we need here is a channel for error propagation and a `defer ctrl.readyFlag.Done()` at the beginning of this function, you don't really want to wait here if `!updated`

Reply:
just looking at the code also, `syncControllerConfig` is also called asynchronously

Reply:
We need to block until we've created the MCs though...e.g. say we error out at 
```
		secret, err := ctrl.kubeClient.CoreV1().Secrets(cfg.Spec.PullSecret.Namespace).Get(cfg.Spec.PullSecret.Name, metav1.GetOptions{})
		if err != nil {
			return ctrl.syncFailingStatus(cfg, err)
		}
```

Then we'll be signaling ready but not have generated the MC.

Reply:
also, if this is something which we need to wait only on MCO startup and not on subsequent syncs, we may want to use https://golang.org/pkg/sync/#Once instead to call Done() w/o setting it to nil

Reply:
Though actually we *do* need to signal ready even if we didn't update a MC since otherwise we will deadlock on restart (which I just hit).

Reply:
it is a good match, I'm not questioning that, just the nilness check and set